include/mw/mmr/backends/FileBackend.h

```
@@ -108,74 +71,40 @@ class FileBackend : public IBackend
 
     Leaf GetLeaf(const LeafIndex& idx) const final
     {
-        const uint64_t leafIndex = idx.GetLeafIndex();
-        if (m_pPositionFile != nullptr)
-        {
-            PosEntry posEntry = GetPosEntry(leafIndex);
-            std::vector<uint8_t> data = m_pDataFile->Read(posEntry.position, posEntry.size);
-            return Leaf::Create(idx, std::move(data));
+        mw::Hash hash = GetHash(idx.GetNodeIndex());
+        auto it = m_leafMap.find(hash);
+        if (it != m_leafMap.end()) {
+            return m_leaves[it->second];
         }
-        else
-        {
-            std::vector<uint8_t> data = m_pDataFile->Read(leafIndex * m_fixedLength, m_fixedLength);
-            return Leaf::Create(idx, std::move(data));
+        LeafDB ldb(m_pDatabase.get());
+        auto pLeaf = ldb.Get(idx, std::move(hash));
+        if (!pLeaf) {
+            return Leaf();  // pruned?
```

We should throw an exception if the leaf isn't found.